File: Python/colas/cola.py
from nodos.nodo import Nodo


class Cola[Elemento]:
	def __init__(self) -> None:
		self.__primero: Nodo[Elemento] | None = None
		self.__ultimo: Nodo[Elemento] | None = None
		self.__tamanio = 0

	def encolar(self, elemento: Elemento) -> None:
		nuevo = Nodo(elemento)

		if self.esta_vacia():
			self.__primero = nuevo
			self.__ultimo = nuevo
		else:
			self.__ultimo.set_siguiente(nuevo)
			self.__ultimo = nuevo

		self.__tamanio += 1

	def desencolar(self) -> Elemento | None:
		if self.esta_vacia():
			return None

		elemento = self.__primero.get_elemento()
		self.__primero = self.__primero.get_siguiente()
		self.__tamanio -= 1

		if self.__tamanio == 0:
			self.__ultimo = None

		return elemento

	def consultar_primero(self) -> Elemento | None:
		if self.esta_vacia():
			return None

		return self.__primero.get_elemento()

	def tamanio(self) -> int:
		return self.__tamanio

	def esta_vacia(self) -> bool:
		return self.__primero is None

File: Java Pilas y Colas/colas/Cola.java
// Cola.java
package colas;

import nodos.Nodo;

public class Cola<T> {

    private Nodo<T> primero;
    private Nodo<T> ultimo;
    private int tamanio;

    public Cola() {
        primero = null;
        ultimo = null;
        tamanio = 0;
    }

    public void encolar(T elemento) {
        Nodo<T> nuevo = new Nodo<>(elemento);

        if (estaVacia()) {
            primero = nuevo;
            ultimo = nuevo;
        } else {
            ultimo.setSiguiente(nuevo);
            ultimo = nuevo;
        }

        tamanio++;
    }

    public T desencolar() {
        if (estaVacia()) {
            return null;
        }

        T elemento = primero.getElemento();
        primero = primero.getSiguiente();

        tamanio--;

        if (tamanio == 0) {
            ultimo = null;
        }

        return elemento;
    }

    public T consultarPrimero() {
        if (estaVacia()) {
            return null;
        }

        return primero.getElemento();
    }

    public int tamanio() {
        return tamanio;
    }

    public boolean estaVacia() {
        return primero == null;
    }
}